hnn_core/viz.py

```diff
@@ -1185,3 +1198,247 @@ def plot_laminar_csd(times, data, contact_labels, ax=None, colorbar=True,
+                 voltage_colormap='viridis', elev=10, azim=-500,
+                 xlim=(-200, 3100), ylim=(-200, 3100), zlim=(-300, 2200),
+                 trial_idx=0, time_idx=0):
+        import matplotlib.pyplot as plt
+        from matplotlib import colormaps
+        self.net = net
+        self.times = net.cell_response.times
+
+        self._vmin = vmin
+        self._vmax = vmax
+
+        self._bg_color = bg_color
+        self._voltage_colormap = voltage_colormap
+
+        self.colormaps = colormaps  # Saved for voltage_colormap update method
+        self.colormap = colormaps[voltage_colormap]
+
+        # Axes limits and view positions
+        self._xlim = xlim
+        self._ylim = ylim
+        self._zlim = zlim
+        self._elev = elev
+        self._azim = azim
+
+        self._trial_idx = trial_idx
+        self._time_idx = time_idx
+
+        # Get voltage data and corresponding colors
+        self.vsec_array = self.get_voltages()
+        self.color_array = self.colormap(self.vsec_array)
+
+        # Create figure
+        if ax is None:
+            self.fig = plt.figure()
+            self.ax = self.fig.add_subplot(projection='3d')
+            self.ax.set_facecolor(self._bg_color)
+        else:
+            self.fig = None
+        self.init_network_plot()
+        self._update_axes()
+
+    def get_voltages(self):
+        vsec_list = list()
+        for cell_type in self.net.cell_types:
+            gid_range = self.net.gid_ranges[cell_type]
+            for gid in gid_range:
+
+                cell = self.net.cell_types[cell_type]
+
+                for sec_name in cell.sections.keys():
+                    vsec = np.array(self.net.cell_response.vsec[
+                        self.trial_idx][gid][sec_name])
+                    vsec_list.append(vsec)
+
+        vsec_array = np.vstack(vsec_list)
+        vsec_array = (vsec_array - self.vmin) / (self.vmax - self.vmin)
+        return vsec_array
+
+    def update_section_voltages(self, t_idx):
+        color_list = self.color_array[:, t_idx]
+        for line, color in zip(self.ax.lines, color_list):
+            line.set_color(color)
+
+    def init_network_plot(self):
+        for cell_type in self.net.cell_types:
+            gid_range = self.net.gid_ranges[cell_type]
+            for gid_idx, gid in enumerate(gid_range):
+
+                cell = self.net.cell_types[cell_type]
+
+                pos = self.net.pos_dict[cell_type][gid_idx]
+                pos = (float(pos[0]), float(pos[2]), float(pos[1]))
+
+                cell.plot_morphology(ax=self.ax, show=False,
+                                     pos=pos, xlim=self.xlim,
+                                     ylim=self.ylim, zlim=self.zlim)
+
+    def _update_axes(self):
+        self.ax.set_xlim(self._xlim)
+        self.ax.set_ylim(self._ylim)
+        self.ax.set_zlim(self._zlim)
+
+        self.ax.view_init(self._elev, self._azim)
+
+    def export_movie(self, fname, dpi=300):
```

Comment:
@jasmainak here's the export function!

Reply:
There are some solutions here that might speed up the exporting by circumventing the matplotlib animation module:
https://stackoverflow.com/questions/4092927/generating-movie-from-python-without-saving-individual-frames-to-files

Reply:
Is it possible to add a progressbar using a callback? I don't think the speed is a showstopper here. Maybe a param called "decim" could be added to take every nth frame. Make it work, make it nice, make it fast ...

Reply:
Definitely would be a nice addition but I think I'd need to be much more matplotlib savvy to know how to implement that :smile: